docs/context_locals.rst

```diff
@@ -0,0 +1,375 @@
+            _app_ctx_stack.pop()
+
+        def __enter__(self):
+            self.push()
+            return self
+
+        def __exit__(self, exc_type, exc_value, tb):
+            self.pop()
+
+    class RequestContext(object):
+        ...
+
+        def push(self):
+            # Before we push the request context we have to ensure that there
+            # is an application context.
+            app_ctx = _app_ctx_stack.top
+            if app_ctx is None or app_ctx.app != self.app:
+                app_ctx = self.app.app_context()
+                app_ctx.push()
+                self._implicit_app_ctx_stack.append(app_ctx)
+            else:
+                self._implicit_app_ctx_stack.append(None)
+
+            _request_ctx_stack.push(self)
+
+        def pop(self):
+            _request_ctx_stack.pop()
+
+            app_ctx = self._implicit_app_ctx_stack.pop()
+            if app_ctx is not None:
+                app_ctx.pop()
+
+        def __enter__(self):
+            self.push()
+            return self
+
+        def __exit__(self, exc_type, exc_value, tb):
+            self.pop()
+
+Notice that each context also provides the ``push()`` (which binds it to the
+current context) and ``pop()`` (which does the opposite) methods, which are
+useful for playing in the console::
+
+    >>> from flask import Flask, current_app
+    >>> app = Flask(__name__)
+    >>> ctx = app.app_context()
+    >>> ctx
+    <flask.ctx.AppContext object at 0x110359190>
+    >>> current_app
+    <LocalProxy unbound>
+    >>> ctx.push()
+    >>> current_app
+    <Flask '__main__'>
+    >>> ctx.pop()
+    >>> current_app
+    <LocalProxy unbound>
+
+Finally, we reach the Flask application which simply creates a request context
+for every new request::
+
+    from .ctx import RequestContext
+
+    class Flask(_PackageBoundObject):
+        ...
+        def app_context(self):
+            return AppContext(self)
+
+        def request_context(self, environ):
+            return RequestContext(self, environ)
+
+        def wsgi_app(self, environ, start_response):
+            with self.request_context(environ):
+                try:
+                    response = self.full_dispatch_request()
+                except Exception as e:
+                    response = self.make_response(self.handle_exception(e))
+                return response(environ, start_response)
+
+Footnotes
+--------------------------------------------------------------------------------
+
+.. [*]
+    For example, Django's internalization module inspects the current request to
+    determine the current language is. [2]_ And the database often keeps data
+    around depending on the current transaction. [2]_
+
+.. [1] http://flask.pocoo.org/docs/design/
+
+.. [2]
+    Ronacher. 2011. "Opening the Flask".
+
+    Slides: http://mitsuhiko.pocoo.org/flask-pycon-2011.pdf
+
+    Presentation: http://blip.tv/pycon-us-videos-2009-2010-2011/pycon-2011-opening-the-flask-4896892
```

Comment:
Looks like this link is dead, but I believe this link contains a full presentation video: http://pyvideo.org/pycon-us-2011/pycon-2011--opening-the-flask.html